### PR TITLE
qa/mgr/dashboard: fix test race condition

### PR DESCRIPTION
**Pull request by epuertat**
Fixes: https://tracker.ceph.com/issues/66844

Given the random nature/irreproducibility of this failure, it's hard to know whether this fixed it or not. In case a retry happens, it'll be printed to the Jenkins console logs (I can even create a Jenkins Build Failure parser to detect this condition).

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by epuertat** [comment]
jenkins test api

**Comment by epuertat** [comment]
jenkins test api

**Comment by epuertat** [comment]
jenkins test make check

**Comment by epuertat** [comment]
jenkins test api

**Comment by epuertat** [comment]
jenkins test api

**Comment by epuertat** [comment]
@nizamial09 @cbodley can you plz review it? Thanks!

**Review by nizamial09 — APPROVED** [review]
lgtm, thank you!

File: qa/tasks/mgr/dashboard/helper.py [hunk 1]
```diff
@@ -722,3 +723,25 @@ def _validate_json(val, schema, path=[]):
...
 
     assert False, str(path)
+
+
+def retry(
+        on_exception: Union[Type[Exception], Tuple[Type[Exception], ...]],
+        tries=3,
+        delay=0,
+        logger: Optional[logging.Logger] = None,
+):
```
Comment (nizamial09) [inline qa/tasks/mgr/dashboard/helper.py:733]: can we use this same `retry` for the mgr failover scenario? basically retrying in the case its failing to establish a connection to an mgr.

**Comment by nizamial09** [comment]
its not related to this PR but saw this error in the [dashboard e2e now](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/15907/consoleFull#20918908177dc3600c-28d8-42e2-9653-291a0c275fdf)

from [logs](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/15907/artifact/build/run/2/out/mgr.x.log)
```
2024-08-05T20:34:52.816+0000 7f0c87df7640  0 [dashboard ERROR service] Command '['radosgw-admin', '-c', '/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/build/run/2/ceph.conf', '-k', '/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/build/run/2/dev/mgr.x/keyring', '-n', 'mgr.x', 'user', 'create', '--uid', 'dashboard', '--display-name', 'Ceph Dashboard', '--system']' timed out after 10 seconds
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/dashboard/services/service.py", line 139, in _get_user_keys
    _, out, err = mgr.send_rgwadmin_command(rgw_create_user_cmd)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/mgr_module.py", line 2511, in send_rgwadmin_command
    result = subprocess.run(  # pylint: disable=subprocess-run-check
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1152, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.10/subprocess.py", line 2004, in _communicate
    self._check_timeout(endtime, orig_timeout, stdout, stderr)
  File "/usr/lib/python3.10/subprocess.py", line 1196, in _check_timeout
    raise TimeoutExpired(
subprocess.TimeoutExpired: Command '['radosgw-admin', '-c', '/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/build/run/2/ceph.conf', '-k', '/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/build/run/2/dev/mgr.x/keyring', '-n', 'mgr.x', 'user', 'create', '--uid', 'dashboard', '--display-name', 'Ceph Dashboard', '--system']' timed out after 10 seconds
2024-08-05T20:34:52.816+0000 7f0c87df7640  0 [dashboard ERROR service] 
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/dashboard/services/service.py", line 175, in configure_rgw_credentials
    assert access_key and secret_key
AssertionError
2024-08-05T20:34:52.816+0000 7f0c87df7640 20 mgr ~Gil Destroying new thread state 0x559a6ce80ea0
2024-08-05T20:34:52.816+0000 7f0c87df7640 -1 mgr.server reply reply (22) Invalid argument No RGW credentials found, please consult the documentation on how to enable RGW for the dashboard.
2024-08-05T20:34:52.816+0000 7f0c87df7640  1 -- [v2:172.21.15.15:6838/1287343968,v1:172.21.15.15:6839/1287343968] --> 172.21.15.15:0/2579729563 -- mgr_command_reply(tid 0: -22 No RGW credentials found, please consult the documentation on how to enable RGW for the dashboard.) -- 0x559a6cd5c340 con 0x559a6d022c00
2024-08-05T20:34:52.816+0000 7f0c87df7640 10 mgr.server operator()  command returned -22
```

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
dashboard cephadm e2e failure is unrelated. merging this

File: qa/tasks/mgr/dashboard/helper.py (hunk 1 above)
Comment (epuertat) [reply qa/tasks/mgr/dashboard/helper.py:733]: Yes, we may use this decorator to cover other situations!

**Comment by epuertat** [comment]
Thank you, @nizamial09 !

**Pull request merged**